resteasy-client-jetty/src/main/java/org/jboss/resteasy/client/jaxrs/engines/jetty/JettyClientEngine.java

```diff
@@ -112,112 +97,75 @@ public <T> CompletableFuture<T> submit(ClientInvocation request, boolean buffere
-                if (callback != null) {
-                    callback.failed(e);
+            final OutputStreamRequestContent contentOut = new OutputStreamRequestContent(
+                    invocation.getHeaders().getMediaType().toString());
+            asyncExecutor.execute(() -> {
+                try {
+                    invocation.writeRequestBody(contentOut.getOutputStream());
+                } catch (Exception e) {
+                    future.completeExceptionally(e);
+                    if (callback != null) {
+                        callback.failed(e);
+                    }
+                } finally {
+                    contentOut.close();
                 }
-                return future;
-            }
-            request.content(new BytesContentProvider(os.toByteArray()));
+            });
+            request.body(contentOut);
         }
 
-        request.send(new Response.Listener.Adapter() {
+        request.send(new InputStreamResponseListener() {
             private ClientResponse cr;
-            private JettyResponseStream stream = new JettyResponseStream();
 
             @Override
+            @SuppressWarnings("unchecked")
             public void onHeaders(Response response) {
-                cr = new JettyClientResponse(invocation.getClientConfiguration(), stream, () -> {
-                    if (!completing.get()) {
-                        future.cancel(true);
-                    }
-                });
+                super.onHeaders(response);
+                InputStream inputStream = getInputStream();
+                cr = new JettyClientResponse(invocation.getClientConfiguration(), inputStream);
                 cr.setProperties(invocation.getMutableProperties());
                 cr.setStatus(response.getStatus());
                 cr.setHeaders(extract(response.getHeaders()));
-                if (!buffered) {
-                    complete();
-                }
-            }
-
-            @Override
-            public void onContent(Response response, ByteBuffer buf) {
-                final ByteBufferPool bufs = client.getByteBufferPool();
-                final ByteBuffer copy = bufs.acquire(buf.remaining(), false);
-                copy.limit(buf.remaining());
-                copy.put(buf);
-                copy.flip();
-                stream.offer(copy, new ReleaseCallback(bufs, copy));
-            }
-
-            @Override
-            public void onSuccess(Response response) {
-                if (buffered) {
+                asyncExecutor.submit(() -> {
                     try {
-                        complete();
+                        if (buffered) {
+                            cr.bufferEntity();
+                        }
+                        complete(extractor == null ? (T) cr : extractor.extractResult(cr));
                     } catch (Exception e) {
+                        try {
+                            inputStream.close();
```

Comment:
It looks like the stream is only ever closed if there is an error. Shouldn't it also be closed on complete? Or at least after it's buffered.

Reply:
I think it does not matter much - all the close is doing is releasing any buffers it still has. On success, there will be no buffers left. But you're right, let's be defensive in case the impl changes later.

Reply:
Oh, I see why this is this way. It is entirely allowable to call `readEntity(InputStream.class)` or otherwise process the data in a streaming manner - so we cannot close the input stream here, otherwise the client cannot access the streamed data in the ClientResponse.

Reply:
I added a close after buffering the entity, since in that case we won't need the stream. But otherwise I think we have to leave it open here.

Reply:
And, in the other case - where the engine is not buffering - the close is taken care of by JettyClientResponse->BuiltResponse.releaseConnection-> `is.close()`.

Reply:
Oh, closing it after buffering is redundant too. the `bufferEntity` implementation already calls `releaseConnection` for us.

Reply:
Ah, right. Sorry, you're right.